### PR TITLE
Remove most of the boilerplate of defining ops with assemblyFormat

### PR DESCRIPTION
**Pull request by krzysz00**
Instead of defining parsers and printers in MIOpenOps.cpp, we use the
assemblyFormat mechanism in MLIR tablegen to specify the format
declaritively.

This has lead to some breaking changes, mainly around ops that take
variadic arguments, which now can't rely on the implicit equalities
between the variadic input and output we were imposing (adding back
support for that is something we might want to give upstream).
Similarly, the types of variadic arguments are no longer inferred.

This changes the printing format for:
  - move_pos_v2
  - threadwise_copy
  - threadwise_copy_v2
  - blockwise_gemm_v2
  - xdlops_gemm_v2
  - threawise_load
  - threadwise_store
and so the tests have been adjusted accordingly

Similarly, fill now needs to give the type of both its
arguments (supporting that equality would be a nice addition tho MLIR,
but may not be worth the bother if we move to linalg.fill).

While I was accounting for these changes, I also fixed some extremely
long lines in the tests.

### CODE REVIEW

File: mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td [hunk 1]
```diff
@@ -16,6 +16,12 @@
 include "mlir/IR/OpBase.td"
 include "mlir/Interfaces/SideEffectInterfaces.td"
 
+// TODO: upstream this
+class Fixed1DVector<int length, Type type> :
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td]: @krzysz00 could you explain the merit adding this class?

File: mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td [hunk 2]
```diff
@@ -206,15 +246,16 @@ def MIOpen_WorkitemIdOp:
...
 }
 
 // blockwise_copy
 def MIOpen_BlockwiseCopyOp:
     MIOpen_Op<"blockwise_copy">,
     Arguments<(ins AnyMemRef:$source,
                    AnyMemRef:$dest,
-                   VectorOfLengthAndType<[3], [I32]>:$sourceCoordVector,
-                   VectorOfLengthAndType<[3], [I32]>:$destCoordVector,
+                   Fixed1DVector<3, I32>:$sourceCoordVector,
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td]: same as my question above, I don't quite see the merit of going from `VectorOfLengthAndType` to `Fixed1DVector` as of yet. could you help share your insights here?

File: mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td [hunk 3]
```diff
@@ -150,28 +177,38 @@ def MIOpen_SubviewOp:
...
-    MIOpen_Op<"fill">,
+    MIOpen_Op<"fill", [AllElementTypesMatch<["input", "value"]>]>,
     Arguments<(ins AnyMemRef:$input,
                    AnyTypeOf<[AnyInteger, AnyFloat]>:$value)> {
   let summary = "Fill memory with constant value on GPU";
   let description = [{
     The `miopen.fill` op fills a memref on GPU with a constant value.
   }];
+  let assemblyFormat = [{
+    `(` operands `)` attr-dict `:` type(operands)
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:189]: from the updated unit test, the syntax feels redundant to me.
for a memref filled with `f32` we should expect the constant value be of the same type `f32`.

File: mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td [hunk 4]
```diff
@@ -150,28 +177,38 @@ def MIOpen_SubviewOp:
...
+      AllTypesMatch<["input", "output"]>]>,
     Arguments<(ins VectorOfRankAndType<[1], [AnyInteger, AnyFloat]>:$input,
                    Variadic<AnyTypeOf<[AnyInteger, AnyFloat]>>:$values)>,
     Results<(outs VectorOfRankAndType<[1], [AnyInteger, AnyFloat]>:$output)> {
     let summary = "Add values to indices within the vector.";
     let description = [{
       The `miopen.move_pos_v2` op adds values to indices within the vector.
     }];
+    let assemblyFormat = [{
+      `(` operands `)` attr-dict `:` type($input) `,` type($values)
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:204]: same as above, looking at the updated unit test, I feel the syntax be a bit redundant to me.
for a vector of `i32` we should expect the offsets also be of type `i32`.

File: mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td [hunk 5]
```diff
@@ -292,6 +347,9 @@ def MIOpen_ThreadwiseLoadOp:
     allowed:
     - Global (generic tensor) to register (naive tensor).
   }];
+  let assemblyFormat = [{
+    `(` operands `)` attr-dict `:` type($source) `,` type($sourceCoord) `->` type($result)
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:345]: it's nice to have the type of the result be printed in the assembly. thanks!

File: mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td [hunk 6]
```diff
@@ -361,6 +429,9 @@ def MIOpen_BlockwiseGemmV2Op:
     - Matrix A and Matrix B shall reside on LDS (naive tensor).
     - Matrix C shall be vectors.
   }];
+  let assemblyFormat = [{
+    `(` operands `)` attr-dict `:` type(operands) `->` type($vectorDs)
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:427]: it's nice to have the type of the result be printed in the assembly. thanks!

File: mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td [hunk 7]
```diff
@@ -400,12 +478,15 @@ def MIOpen_XdlopsGemmV2Op:
...
                    Variadic<VectorOfRankAndType<[1], [F32, F16]>>:$vectorCs)>,
-    Results<(outs Variadic<VectorOfRankAndType<[1], [F32, F16]>>: $vectorDs)> {
+    Results<(outs Variadic<VectorOfRankAndType<[1], [F32, F16]>>:$vectorDs)> {
   let summary = "XDLOPS GEMM V2";
   let description = [{
     The `miopen.xdlops_gemm_v2` op is an abstraction of doing GEMM based on XDLOPS.
     It would employ a series of `miopen.mfma` operations.
   }];
+  let assemblyFormat = [{
+    `(` operands `)` attr-dict `:` type(operands) `->` type(results)
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:482]: it's nice to have the type of the result be printed in the assembly. thanks!

**Review by whchung — CHANGES_REQUESTED** [review]
It's nice to migrate to declarative way to specify the syntax. The less code we write the better.

On the other hand, I'm having some different opinions on some of the proposed breaking syntax changes. Please help take a look at my comments and we can iterate from there.

File: mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td (hunk 1 above)
Comment (krzysz00) [reply mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td]: The reason this has been added is to allow us to not have to write the type of the vector when we know it's length and size - hence the `SameBuildabilityAs`. We used this inferred vector type in a few places in our formats.

File: mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td (hunk 2 above)
Comment (krzysz00) [reply mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td]: To rephrase, having Fixed1DVector lets us not have a `type($sourceCoordVector)` in the format

File: mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td (hunk 3 above)
Comment (krzysz00) [reply mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:189]: Agreed, though the declaritive syntax doesn't currently have a mechanism for imposing that constraint.

We could go back to a custom parser/printer for this one, or live with the redudancy for now and send in a patch to support this type of reasoning.

File: mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td (hunk 4 above)
Comment (krzysz00) [reply mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:204]: Same equality issue as above, as well as the fact that the vector type is less constrained than the offsets.

File: mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td (hunk 5 above)
Comment (krzysz00) [reply mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:345]: No problem!

**Comment by krzysz00** [comment]
I agree, overall, with your complaints about some of the breaking changes. The alternatives for this PR are:

1. Restore the custom code for the redundancies (`fill` and `move_pos`), and then fix things later
2. Live with it for now

In any case, what I'd want to see eventually is support for
1. `AllElementTypesMatch` allowing the MLIR code to resolve the type of redudancy in "Fill"
2. Some sort of shorthand for variadic operands of fixed type
both of which are changes to be tossed at MLIR proper

**Comment by whchung** [comment]
> I agree, overall, with your complaints about some of the breaking changes. The alternatives for this PR are:
> 
> 1. Restore the custom code for the redundancies (`fill` and `move_pos`), and then fix things later
> 2. Live with it for now

I'm all for removing boilerplate codes, so let's live with it.

**Comment by krzysz00** [comment]
> I'm all for removing boilerplate codes, so let's live with it.

That's what I was thinking too, but I'm willing to be persuaded otherwise if folks have objections to some of these changes.

There other things you're still unhappy about?

File: mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td (hunk 2 above)
Comment (whchung) [reply mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td]: I'm actually working on changing how we use `blockwise_copy` so there could be cases we pass in 3D vectors or 4D vectors. Having this `Fixed1DVector` seems to remove such flexibility offered by `VectorOfLengthAndType`.

**Review by whchung — CHANGES_REQUESTED** [review]
I'm okay with the breaking changes for some ops, but I'm holding reservation to `Fixed1DVector`.

**Comment by krzysz00** [comment]
So, the reason I have Fixed1DVector is that the existing format for those ops didn't print the vector type in question, and assumed it was always `vector<3xi32>`. I had to add a new class to allow for those types to be inferred automatically.

If you instead want to explicitly add the type of the coordinate vectors to the IR (which'll support your upcoming change), I can do that instead.

**Comment by whchung** [comment]
> If you instead want to explicitly add the type of the coordinate vectors to the IR (which'll support your upcoming change), I can do that instead.

SGTM. Thanks!

**Comment by krzysz00** [comment]
@whchung  - the coordinate vectors have been explicitly added to the output of the blockwise operations. Hopefully this makes your upcoming change easier.

**Review by whchung — APPROVED** [review]

**Pull request merged**